modules/failure-handler/src/test/java/org/apache/ignite/internal/failure/FailureProcessorThreadDumpThrottlingTest.java

```diff
@@ -0,0 +1,232 @@
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+
+                // Trigger the failure processing twice.
+                for (int i = 0; i < 2; i++) {
+                    assertThat(failureProcessor.process(failureCtx), is(false));
+                }
+            });
+        } finally {
+            logInspector.stop();
+        }
+
+        assertThat(threadDumpMessageCounter.get(), is(2));
+        assertThat(throttlingMessageCounter.get(), is(2));
+    }
+
+    /**
+     * Tests that thread dumps will be throttled per failure type and will be generated again after timeout exceeded.
+     */
+    @Test
+    public void testThrottlingPerFailureType() {
+        AtomicInteger threadDumpMessageCounter = new AtomicInteger();
+        AtomicInteger throttlingMessageCounter = new AtomicInteger();
+
+        LogInspector logInspector = new LogInspector(FailureManager.class.getName());
+
+        logInspector.addHandler(
+                evt -> evt.getMessage().getFormattedMessage().startsWith(THREAD_DUMP_MSG),
+                threadDumpMessageCounter::incrementAndGet);
+        logInspector.addHandler(
+                evt -> evt.getMessage().getFormattedMessage().startsWith("Thread dump is hidden due to throttling settings."),
+                throttlingMessageCounter::incrementAndGet);
+
+        logInspector.start();
+        try {
+            testFailureProcessing(enabledThreadDumpWithThrottlingConfiguration, failureProcessor -> {
+                FailureContext workerBlockedFailureCtx =
+                        new FailureContext(SYSTEM_WORKER_BLOCKED, new Throwable("Failure context error"));
+
+                FailureContext opTimeoutFailureCtx =
+                        new FailureContext(SYSTEM_CRITICAL_OPERATION_TIMEOUT, new Throwable("Failure context error"));
+
+                for (int i = 0; i < 2; i++) {
+                    failureProcessor.process(workerBlockedFailureCtx);
+                    failureProcessor.process(opTimeoutFailureCtx);
+                }
+
+                try {
+                    Thread.sleep(3000);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+
+                for (int i = 0; i < 2; i++) {
+                    failureProcessor.process(workerBlockedFailureCtx);
+                    failureProcessor.process(opTimeoutFailureCtx);
+                }
+            });
+        } finally {
+            logInspector.stop();
+        }
+
+        assertThat(threadDumpMessageCounter.get(), is(4));
+        assertThat(throttlingMessageCounter.get(), is(4));
+    }
+
+    /**
+     * Creates a new instance of {@link FailureManager} with the given configuration and runs the test represented by {@code test} closure.
+     */
+    public static void testFailureProcessing(FailureProcessorConfiguration configuration, Consumer<FailureProcessor> test) {
```

Comment:
It is very unusual to have a public method that is not test in a test class. Does it need to be public?

Reply:
Yep, it can be package-private. Changed